src/mono/mono/mini/mini.h

```diff
@@ -2189,7 +2189,14 @@ GString  *mono_print_ins_index_strbuf       (int i, MonoInst *ins);
 void      mono_print_ins                    (MonoInst *ins);
 void      mono_print_bb                     (MonoBasicBlock *bb, const char *msg);
 void      mono_print_code                   (MonoCompile *cfg, const char *msg);
+
+#ifndef DISABLE_LOGGING
 const char* mono_inst_name (int op);
+#else
+#define mono_inst_name(op) _static_assert_disabled_logging(op)
+void _static_assert_disabled_logging(int op) __attribute__((error("mono_inst_name is not available when DISABLE_LOGGING is defined.")));
+#endif
```

Comment:
This is an interesting idea.  We don't only compile with gcc and clang, so we need to make some macros to make it workable.  Also it's better to use `__attribute__((unavailable))` when the compiler supports it, because then optimizations that remove a call won't remove the warning.

I did an [example on godbolt](https://godbolt.org/z/G48js6eYj) that seems to work even with old versions of gcc and clang.

Add this to `src/mono/mono/utils/mono-compiler.h`

```c

#if defined(__clang__)
#define MONO_FUNCTION_UNAVAILABLE(msg) __attribute__((unavailable(msg)))
#elif (__GNUC__ == 12 && __GNUC_MINOR__ >= 1)
#define MONO_FUNCTION_UNAVAILABLE(msg) __attribute__((unavailable(msg)))
#elif (__GNUC__ > 4) || (__GNUC__ >=4 && __GNUC_MINOR__ >= 4)
/* this might not warn if optimization removes a call, but it's better than nothing.
 * disable inlining to encourage GCC not to eliminate the call
 */
#define MONO_FUNCTION_UNAVAILABLE(msg) __attribute__((noinline,error(msg)))
#else
#define MONO_FUNCTION_UNAVAILABLE(msg) /* empty */
#endif
```

And then do this in this header (the implementation in `helpers.c` should be entirely within the `#ifdef` with no `#else` part):

```c
#ifndef DISABLE_LOGGING
const char* mono_inst_name (int op);
#else
MONO_FUNCTION_UNAVAILABLE("mono_inst_name is not available when DISABLE_LOGGING is defined") const char *mono_inst_name (int op);
#endif
```